packages/core/src/stateUtils.ts

```diff
@@ -1475,166 +1091,298 @@ export function microstep<
+    }
+  }
 }
 
 export function resolveMicroTransition<
   TContext extends MachineContext,
   TEvent extends EventObject
 >(
-  machine: StateMachine<TContext, TEvent, any, any, any>,
   transitions: Transitions<TContext, TEvent>,
   currentState: State<TContext, TEvent, any>,
-  predictableExec?: PredictableActionArgumentsExec,
+  actorCtx: ActorContext<any, any> | undefined,
   _event: SCXML.Event<TEvent> = initEvent as SCXML.Event<TEvent>
 ): State<TContext, TEvent, any> {
+  const { machine } = currentState;
   // Transition will "apply" if:
   // - the state node is the initial state (there is no current state)
   // - OR there are transitions
-  // - OR there are eventless transitions (if there are no transitions)
-  const willTransition =
-    currentState._initial ||
-    transitions.length > 0 ||
-    selectEventlessTransitions(currentState, machine).length > 0;
+  const willTransition = currentState._initial || transitions.length > 0;
 
-  const prevConfig = getConfiguration<TContext, TEvent>(
+  const prevConfiguration = getConfiguration(
     !currentState._initial ? currentState.configuration : [machine.root]
   );
 
-  const resolved = microstep<TContext, TEvent>(
-    !currentState._initial
-      ? transitions
-      : [
+  if (!currentState._initial && !willTransition) {
+    const inertState = currentState.clone({
+      _event,
+      actions: [],
+      transitions: []
+    });
+
+    inertState.changed = false;
+    return inertState;
+  }
+
+  const microstate = microstep<TContext, TEvent>(
+    currentState._initial
+      ? [
           {
-            target: [...prevConfig].filter(isAtomicStateNode),
+            target: [...prevConfiguration].filter(isAtomicStateNode),
             source: machine.root,
             actions: [],
             eventType: null as any,
             toJSON: null as any // TODO: fix
           }
-        ],
+        ]
+      : transitions,
     currentState,
-    currentState.context,
-    new Set(prevConfig),
-    machine,
+    prevConfiguration,
     _event,
-    predictableExec
+    actorCtx
   );
 
-  if (!currentState._initial && !willTransition) {
-    const inertState = State.inert(currentState);
-    inertState.event = _event.data;
-    inertState._event = _event;
-    inertState.changed = _event.name === actionTypes.update;
-    return inertState as any;
-  }
-
-  const resolvedConfiguration = willTransition
-    ? Array.from(resolved.configuration)
-    : !currentState._initial
-    ? currentState.configuration
-    : [];
-
-  const meta = resolvedConfiguration.reduce((acc, subStateNode) => {
-    if (subStateNode.meta !== undefined) {
-      acc[subStateNode.id] = subStateNode.meta;
-    }
-    return acc;
-  }, {} as Record<string, string>);
-
-  const { context, actions: nonRaisedActions } = resolved;
+  const { context, actions: nonRaisedActions } = microstate;
 
   const children = { ...currentState.children };
-  setChildren(children, nonRaisedActions);
+  setChildren();
```

Comment:
Similarly here - with explicit arguments, i feel like it's easier to follow the control flow. When looking at this statement alone I'm like "what kind of magic this is?"